File: Dynamic-program/474-findMaxForm/474.py
class Solution(object):
    def findMaxForm(self, strs, m, n):
        """
        :type strs: List[str]
        :type m: int
        :type n: int
        :rtype: int
        """
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for str in strs:
            zero_count = str.count('0')
            one_count = str.count('1')
            for i in range(m, zero_count - 1, -1):
                for j in range(n, one_count - 1, -1):
                    dp[i][j] = max(dp[i][j], dp[i - zero_count][j - one_count] + 1)
        return dp[m][n]
    
obj = Solution()
print(obj.findMaxForm(["10", "0001","111001","1","0"], 5, 3))

File: Dynamic-program/474-findMaxForm/474.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int findMaxForm(vector<string> &strs, int m, int n)
    {
        // dp[i][j]: 最多有i个0和j个1的子集的最大长度
        // 对于新的二进制字符串k，其中0和1数量为k_0和k_1:
        // dp[i][j] = max(dp[i][j], dp[i - k_0][j - k_1] + 1)
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        for (const string &str : strs) {
            int zeroCount = count(str.begin(), str.end(), '0');
            int oneCount = count(str.begin(), str.end(), '1');
            for (int i = m; i >= zeroCount; --i) {
                for (int j = n; j >= oneCount; --j) 
                    dp[i][j] = max(dp[i][j], dp[i - zeroCount][j - oneCount] + 1);
            }
        }
        return dp[m][n];
    }
};